org.metaborg.spoofax.core/src/main/java/org/metaborg/spoofax/core/completion/JSGLRCompletionService.java

```diff
@@ -2,84 +2,752 @@
+                            logger.error("Unexpected proposal term {}, skipping", proposalTerm);
+                            continue;
+                        }
+
+                        ITokenizer tokenizer = ImploderAttachment.getTokenizer(oldNode);
+                        final ImploderAttachment oldNodeIA = oldNode.getAttachment(ImploderAttachment.TYPE);
+
+                        // get the last non-layout token before the new node
+                        int tokenPosition =
+                            oldNodeIA.getLeftToken().getIndex() - 1 > 0 ? oldNodeIA.getLeftToken().getIndex() - 1 : 0;
+                        while((tokenizer.getTokenAt(tokenPosition).getKind() == IToken.TK_LAYOUT || tokenizer
+                            .getTokenAt(tokenPosition).getKind() == IToken.TK_ERROR) && tokenPosition > 0)
+                            tokenPosition--;
+
+                        insertionPoint = tokenizer.getTokenAt(tokenPosition).getEndOffset();
+
+                        // skip all the (erroneous) characters that were in the text already
+                        suffixPoint = oldNodeIA.getRightToken().getEndOffset() + 1;
+                    }
+
+                    final Completion completion = new Completion(name, description, insertionPoint + 1, suffixPoint);
+                    completions.add(completion);
+                }
+            }
+        }
+        return completions;
+
+    }
+
+    public Collection<ICompletion> listsCompletions(int position, Iterable<IStrategoList> lists, String input,
+        ILanguageImpl language, FileObject location) throws MetaborgException {
+
+        Collection<ICompletion> completions = Lists.newLinkedList();
+        int insertionPoint = position;
+        int suffixPoint = insertionPoint + 1;
+
+
+        for(ILanguageComponent component : language.components()) {
+            final ITermFactory termFactory = termFactoryService.get(component);
+
+            for(IStrategoList list : lists) {
+                ListImploderAttachment attachment = list.getAttachment(null);
+                String placeholderName =
+                    attachment.getSort().substring(0, attachment.getSort().length() - 1) + "-Plhdr";
+                IStrategoAppl listPlaceholder = termFactory.makeAppl(termFactory.makeConstructor(placeholderName, 0));
+                final IStrategoTerm strategoInput =
+                    termFactory.makeTuple(list, listPlaceholder, termFactory.makeInt(position));
+                final HybridInterpreter runtime = strategoRuntimeService.runtime(component, location);
+                final IStrategoTerm proposalsLists =
+                    strategoCommon.invoke(runtime, strategoInput, "get-proposals-list");
+                if(proposalsLists == null) {
+                    logger.error("Getting proposals for {} failed", strategoInput);
+                }
+                for(IStrategoTerm proposalTerm : proposalsLists) {
```

Comment:
Possible NPE here, should `continue` if `proposalsLists` is null, not just log an error.